### PR TITLE
feat: initial set of rules

### PR DESCRIPTION
**Pull request by Averethel**
The goal is to create a set of linter rules that is possibly close to our coding style and actively helps us when needed.

Extension of/changes to this set will be tackled in a separate tech 20 ticket.

TODO:
- [x] initial set of rules
- [x] changes based on dealer-hub
- [x] changes based on listings
- [x] changes based on components
- [x] split rules to:
  - plain js/ts
  - react specific rules
  - nextjs specific rules
- [x] some basic tests
- [x] document the package
- [x] figure out best way of sharing prettier config

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,17 +18,53 @@
   "author": "CAR FOR YOU",
   "license": "MIT",
   "homepage": "https://github.com/carforyou/carforyou-eslint-rules-pkg#readme",
+  "peerDependencies": {
```
Comment (lkappeler) [inline package.json:21]: What was your basis to elaborate on what plugins we wanna use?

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        "jest/expect-expect": "off",
+        "jest/valid-expect-in-promise": "off",
+        "jest/valid-expect": "off",
+        "jest/no-standalone-expect": "off",
+        "testing-library/await-async-query": "off",
+      },
+    },
+    {
+      files: [
+        "**/__tests__/next/**/*.test.tsx",
```
Comment (lkappeler) [inline src/index.js:67]: why not more generic `**/**/*.test.*`

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        "**/__tests__/pages/**/*.test.tsx",
+        "**/__tests__/pages/**/*.test.jsx",
+        "**/components/**/*.test.tsx",
+        "**/components/**/*.test.jsx",
+      ],
+      rules: {
+        // off because of testing library
```
Comment (lkappeler) [inline src/index.js:74]: I can not follow the comment.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        "jest.config.js",
+        "release.config.js",
+        ".eslintrc.js",
+        "*.json",
+      ],
+      parser: "esprima",
+    },
+  ],
+  rules: {
+    "@typescript-eslint/explicit-function-return-type": "off",
```
Comment (lkappeler) [inline src/index.js:91]: The rule sounds useful to me why do you want to disable it?
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/explicit-function-return-type.md

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    "@typescript-eslint/no-unused-vars": [
+      "error",
+      {
+        varsIgnorePattern: "^_",
+        argsIgnorePattern: "^_",
+        ignoreRestSiblings: true,
+      },
+    ],
+    "react/prop-types": "off",
```
Comment (lkappeler) [inline src/index.js:100]: Reason?
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    "react/display-name": "off",
+    "jest/no-mocks-import": "off",
```
Comment (lkappeler) [inline src/index.js:102]: Might be worth the evaluation if we ar using the mocks not as inteded.

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    "prettier/prettier": ["error", { semi: false }],
```
Comment (lkappeler) [inline src/index.js:103]: Reason to override the stadard?

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    "prefer-spread": "off",
```
Comment (lkappeler) [inline src/index.js:104]: Reason?

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    "no-console": ["error", { allow: ["info", "warn", "error"] }],
+    "no-irregular-whitespace": [
```
Comment (lkappeler) [inline src/index.js:106]: Not sure if we should enable it and explicitly disable it when intended

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      "error",
+      {
+        skipStrings: true,
+        skipTemplates: true,
+      },
+    ],
+    "no-restricted-imports": [
```
Comment (lkappeler) [inline src/index.js:113]: I think those should be project-specific not in the linter package as not all project are next projets

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            name: "next/router",
+            importNames: ["default"],
+            message: 'Please import from "~/server/routing/routes.ts" instead.',
+          },
+        ],
+      },
+    ],
+    // TODO: we had that enabled but also disabled on in multiple files
+    "sort-keys": "off",
+    "import/no-internal-modules": [
```
Comment (lkappeler) [inline src/index.js:131]: Similar to the restricted imports I think the project should define those

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    "@typescript-eslint/no-unused-vars": [
+      "error",
+      {
+        varsIgnorePattern: "^_",
+        argsIgnorePattern: "^_",
+        ignoreRestSiblings: true,
+      },
+    ],
+    "react/prop-types": "off",
+    "react/display-name": "off",
```
Comment (lkappeler) [inline src/index.js:101]: No strong opinion but seems useful
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+          "**/__fixtures__/*",
+          "@testing-library/**/*",
+          "next/*",
+          "next-i18next/*",
+          "react-dom/*",
+        ],
+      },
+    ],
+    "json/*": ["error", "allowComments"],
+    "react-hooks/exhaustive-deps": "off",
```
Comment (lkappeler) [inline src/index.js:148]: Found this https://reactjs.org/docs/hooks-rules.html
But did not fully understand what it means, can you explain it a bit

File: package.json (hunk 1 above)
Comment (Averethel) [reply package.json:21]: I picked the plugins based on the tooling we use. Most of the extra plugins do not impact our current code by a lot but help us with some minor problems. E.g. I discovered a bunch of tests that were using RTL helpers that were asynchronous in a synchronous way and vice versa.

File: src/index.js (hunk 2 above)
Comment (Averethel) [reply src/index.js:67]: Because I only want to affect test files that are using react-testing-library

File: src/index.js (hunk 3 above)
Comment (Averethel) [reply src/index.js:74]: Those two linter rules ensure that:
- your test always has an `expect`
- an expect in your test always have a condition

With the testing library, this is no longer the case since testing for the existence of an element can simply use matcher helpers without an expect at all.

File: src/index.js (hunk 4 above)
Comment (Averethel) [reply src/index.js:91]: I agree this can be valuable but we're simply not writing code this way ATM. It's a good candidate for ruleset discussion.

File: src/index.js (hunk 5 above)
Comment (Averethel) [reply src/index.js:100]: We're not using prop-types, we use typescript.
If you write
```
interface Props {
  initialValue: Number
}

const Counter: FC<Props> = ({ initialValue }) => { ... }
```
this plugin will complain, even thought that through declaring `FC<Props>` you're providing type information.

File: src/index.js (hunk 12 above)
Comment (Averethel) [reply src/index.js:101]: Another good candidate for rules discussion, we're simply never had this enabled.

File: src/index.js (hunk 6 above)
Comment (Averethel) [reply src/index.js:102]: We have a bunch of mocked network responses that we import directly. But it's probably not worth disabling the rule....

File: src/index.js (hunk 7 above)
Comment (Averethel) [reply src/index.js:103]: Current configuration:
https://github.com/carforyou/carforyou-tslint-rules-pkg/blob/master/carforyou-tslint-rules.json#L18

File: src/index.js (hunk 9 above)
Comment (Averethel) [reply src/index.js:106]: We want this because we oftentimes include the non-breaking spaces in template strings and translations.

File: src/index.js (hunk 8 above)
Comment (Averethel) [reply src/index.js:104]: current coding style

File: src/index.js (hunk 13 above)
Comment (Averethel) [reply src/index.js:148]: It's a relatively new rule that wants you to declare all variables you reference inside `useEffect` in the dependencies array. We have a bunch of useEffects that depend on `id` only even though they reference the full object and it would force us to add the object to the dependency array making the effect always trigger.

File: src/index.js (hunk 10 above)
Comment (Averethel) [reply src/index.js:113]: Good point. With eslint we can expose multiple configurations so, we can expose a next variant that includes those.

File: src/index.js (hunk 11 above)
Comment (Averethel) [reply src/index.js:131]: That would make this configuration repeated in many projects. I think it's good to have a starting point.
But as mentioned before it makes sense to move it to next/react specific configs.

File: src/index.js (hunk 5 above)
Comment (lkappeler) [reply src/index.js:100]: Yeah I misread that rule...

File: src/index.js (hunk 8 above)
Comment (lkappeler) [reply src/index.js:104]: I have seen very mixed usage so far, but let's discuss it with the other rules...

File: src/index.js (hunk 7 above)
Comment (lkappeler) [reply src/index.js:103]: So what was the reason to override it in the current configuration then? ;)
Let's move it to the discussion...

File: src/index.js (hunk 9 above)
Comment (lkappeler) [reply src/index.js:106]: Maybe I misunderstood the rule but isn't it about linting if you do multiple spaces?

File: src/index.js (hunk 13 above)
Comment (lkappeler) [reply src/index.js:148]: So it's like where we use the deepComare function to trigger the hook for changes deeper in the object?

File: src/index.js (hunk 9 above)
Comment (Averethel) [reply src/index.js:106]: No. It's about using the same character for spacing. So it will alert you if you use `tab` and `spaces` for example.

File: src/index.js (hunk 13 above)
Comment (Averethel) [reply src/index.js:148]: Kind of, but also not exactly. Take this snipper for instance:
```
useEffect(() => {
    const updateProfile = async () => {
      const newProfile = await fetchDealerProfile(activeDealerId)
      if (deepEqual(newProfile, profile)) {
        setDealerProfile(newProfile)
      }
    }

    if (activeDealerId && hasRole(currentUser, Role.PrivateSeller)) {
      updateProfile()
    }
  }, [activeDealerId])
```

what the rule would suggest is to add:
`profile` and `currentUser` to the dependency array.
1. This would require us to switch to deep compare since we're using objects
2. In this particular case was want to only depend on the ID to re-etch data

Comment (lkappeler) [reply src/index.js:148]: Thanks for the explanation 😊, I see how the rule can help with not missing cases/forces you to think about what triggers change the result of you state. Im fine with initialy exlcuding it, let's consider looking in to it with the team.

**Comment by Averethel** [comment]
@lkappeler Yes that's correct. Sorry for not reflecting that in the description.

File: src/index.js [hunk 14]
```diff
@@ -97,6 +97,32 @@ module.exports = {
         ignoreRestSiblings: true,
       },
     ],
+    "@typescript-eslint/camelcase": "off",
+    "@typescript-eslint/naming-convention": [
```
Comment (lkappeler) [inline src/index.js]: Can you summarize what it does, it's quite time intentes to get everything together from the docs

File: src/index.js [hunk 15]
```diff
@@ -123,6 +123,10 @@ module.exports = {
         format: ["PascalCase"],
       },
     ],
+    "@typescript-eslint/no-use-before-define": [
```
Comment (lkappeler) [inline src/index.js]: Let's add that to the list for the discussion

File: src/index.js (hunk 14 above)
Comment (Averethel) [reply src/index.js]: Will mention that in the docs for sure, but here's a short summary:
- a default name for everything is in `camelCase`
- if it's a member or a variable we also accept `PascalCase` - this is for classes/components
- for the properties, we also allow `snake_case` - this covers a bunch of external properties that are expected to be in `snake_case` as well as 99.9% of our BEM convention that we use with classnames
- for the types, enums and interfaces we want to use `PascalCase` only

Comment (lkappeler) [reply src/index.js]: Thanks, sound good to me.

File: .eslintrc.js [hunk 16]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  extends: ["./index.js"],
+  overrides: [
+    {
+      files: ["packMain.js"],
+      parser: "esprima",
```
Comment (masone) [inline .eslintrc.js:6]: what is this custom parser for?

File: index.js [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    "plugin:@typescript-eslint/eslint-recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:prettier/recommended",
+    "prettier/@typescript-eslint",
+    "plugin:jest/recommended",
+    "plugin:json/recommended",
+  ],
+  parser: "@typescript-eslint/parser",
+  parserOptions: {
+    project: "tsconfig.json",
```
Comment (masone) [inline index.js:18]: sometimes we have multiple tsconfigs. sometimes even one for parsing. can these finally go away with this?

File: index.js [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+      rules: {
+        "@typescript-eslint/no-var-requires": "off",
+        "@typescript-eslint/explicit-function-return-type": "off",
+      },
+    },
+    {
+      files: [
+        "babel.config.js",
+        "jest.config.js",
+        "release.config.js",
```
Comment (masone) [inline index.js:44]: `*.config.js` instead?

Comment (masone) [reply index.js:44]: (not sure that works. just inferring from below)

File: index.js [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+        format: null,
+        filter: {
+          regex: "^UNSAFE_",
+          match: true,
+        },
+      },
+      {
+        selector: "typeLike",
+        format: ["PascalCase"],
+      },
```
Comment (masone) [inline index.js:102]: do we really have such a custom setup that we need to specify all these?

File: index.js [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+    ],
+    "@typescript-eslint/no-use-before-define": [
+      "error",
+      { functions: false, variables: false },
+    ],
+    "prettier/prettier": ["error", { semi: false }],
+    "prefer-spread": "off",
+    "no-console": ["error", { allow: ["info", "warn", "error"] }],
```
Comment (masone) [inline index.js:110]: what's the point in still allowing most `console.*` functions?

File: index.js [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+        paths: [
+          {
+            name: "next/link",
+            message: 'Please import from "~/components/appLink.tsx" instead.',
+          },
+          {
+            name: "next/router",
+            importNames: ["default"],
+            message: 'Please import from "~/server/routing/routes.ts" instead.',
+          },
```
Comment (masone) [inline index.js:130]: these are next specific. do you want to move them to the respective config?

Comment (masone) [reply index.js:130]: the message is likely app-specific...

File: index.js [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+        ],
+      },
+    ],
+    // TODO: we had that enabled but also disabled on in multiple files
```
Comment (masone) [inline index.js:134]: TODO

File: next.js [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+      "error",
+      {
+        allow: [
+          "**/*.css",
+          "@testing-library/**/*",
+          "react-dom/*",
+          "**/configuration/**/*",
+          "**/configuration/*",
+          "**/__fixtures__/**/*",
+          "**/__fixtures__/*",
```
Comment (masone) [inline next.js:28]: aren't these cypress-specific? they're not next-specific, right? maybe go for the cypress folder instead?

File: next.js [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+  rules: {
+    "import/no-internal-modules": [
+      "error",
+      {
+        allow: [
+          "**/*.css",
+          "@testing-library/**/*",
+          "react-dom/*",
+          "**/configuration/**/*",
+          "**/configuration/*",
```
Comment (masone) [inline next.js:26]: these are app-specific and for example don't apply to most packages

File: package.json [hunk 25]
```diff
@@ -18,17 +18,52 @@
...
+    "eslint-plugin-cypress": "^2.10.3",
+    "eslint-plugin-import": "^2.20.2",
+    "eslint-plugin-jest": "^23.11.0",
+    "eslint-plugin-jsdoc": "^25.4.1",
+    "eslint-plugin-json": "^2.1.1",
+    "eslint-plugin-prefer-arrow": "^1.2.1",
+    "eslint-plugin-prettier": "^3.1.3",
+    "eslint-plugin-react": "^7.20.0",
+    "eslint-plugin-react-hooks": "^4.0.2",
+    "eslint-plugin-testing-library": "^3.1.2",
```
Comment (masone) [inline package.json:36]: imho, all plusing we actively use should be dependencies. only prettier and eslint should be peer dependencies. I would under no circumstance have to install all of these in the projects

File: package.json [hunk 26]
```diff
@@ -18,17 +18,52 @@
...
+    "prettier": "^2.0.5",
+    "tslint": "^6.1.2"
```
Comment (masone) [inline package.json:38]: likely not required, right?

File: package.json [hunk 27]
```diff
@@ -37,16 +72,20 @@
   "@pika/pack": {
     "pipeline": [
       [
-        "@pika/plugin-ts-standard-pkg",
+        "./packMain.js",
         {
-          "tsconfig": "tsconfig.pika.json"
+          "main": "index.js"
```
Comment (masone) [inline package.json:77]: isn't that what plugin-standard-pkg is for?

Comment (masone) [reply package.json:77]: I think I might be fine with deep imports over forcing this weird main attribute

File: .eslintrc.js (hunk 16 above)
Comment (Averethel) [reply .eslintrc.js:6]: It's not a custom one, it's the default they use for javascript. We use typescript parser, but for files that are not in `tsconfig` you need to switch to a different one otherwise, the typescript parser will detect that the file is not part of the project.

File: index.js (hunk 17 above)
Comment (Averethel) [reply index.js:18]: Yes, the separate config for linting is gone in all refactored projects. We still might have a separate one for pike in package repos, but that's fair game.

File: .eslintrc.js (hunk 16 above)
Comment (masone) [reply .eslintrc.js:6]: Can you make it in a way we don't need to add specific files here as we add them in the project?

File: index.js (hunk 19 above)
Comment (Averethel) [reply index.js:102]: Take a look here, that's my summary of what we cover: https://github.com/carforyou/carforyou-eslint-rules-pkg/pull/1#discussion_r427186803

This rule is not enabled by default, `camelcase` is. But:
1) `camelcase` is deprecated in favour of this one
2) this one actually gives us the flexibility to cover all of our cases and some control if we want to change naming conventions in the future.

File: .eslintrc.js (hunk 16 above)
Comment (Averethel) [reply .eslintrc.js:6]: Not sure I follow. It's a project-specific file that does not fall into the typescript project.
Where would you configure it if not in the project-specific linter setup?

File: index.js (hunk 20 above)
Comment (Averethel) [reply index.js:110]: My rationale was that we use console statements and most of them are justified. We still error on `debug` and `log` which are the ones you use for local debugging.

File: index.js (hunk 21 above)
Comment (Averethel) [reply index.js:130]: Yeah makes sense to move. I like this message since it tells you what you're doing wrong and how to fix it vs. a generic `this import is not allowed`. Do you have a better suggestion?

File: index.js (hunk 22 above)
Comment (Averethel) [reply index.js:134]: It's a rule we want to discuss.

File: index.js (hunk 20 above)
Comment (masone) [reply index.js:110]: I'd personally prefer to whitelist specific cases with a comment over having global ways out

File: next.js (hunk 24 above)
Comment (Averethel) [reply next.js:26]: That's why they are in `nextjs` file.

File: .eslintrc.js (hunk 16 above)
Comment (masone) [reply .eslintrc.js:6]: I mixed up things. This file is only used in the eslint-rules-pkg and extends the configuration we expose. It is not part of the config that the projects will consume

File: package.json (hunk 25 above)
Comment (Averethel) [reply package.json:36]: Yes, I agree. I had some problems with one or two plugins that needed to be installed in the project despite being dependencies of the rules package. Will try to isolate and only include those here.

File: package.json (hunk 26 above)
Comment (Averethel) [reply package.json:38]: There are still some tslint rules that are run by tslint inside eslint. I'll take a closer look at that.

File: package.json (hunk 27 above)
Comment (Averethel) [reply package.json:77]: We don't use standard-pkg here, since it requires the code to be in the `src` directory and `copy-assets` cannot copy from directory to the root of the package. That being said we might not need that since it's `index.js` anyway.

File: index.js (hunk 20 above)
Comment (Averethel) [reply index.js:110]: I like allowing sensible cases and give you a way out if you want to use `console` in a sensible way...

File: next.js (hunk 23 above)
Comment (Averethel) [reply next.js:28]: True, moved to react setup that includes cypress.

File: package.json (hunk 25 above)
Comment (Averethel) [reply package.json]: Turns out it works without any problems by using deeps of the package. No need for peer deps at all. If I'm to guess it was probably hiccups due to initially local development. Will remove unnecessary deps everywhere.

File: index.js [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+    {
+      files: ["*.js"],
+      rules: {
+        "@typescript-eslint/no-var-requires": "off",
+        "@typescript-eslint/explicit-function-return-type": "off",
+      },
+    },
+    {
+      files: ["*.config.js", ".eslintrc.js", "*.json"],
+      parser: "esprima",
```
Comment (Averethel) [inline index.js:43]: let's try to revert - use esprima by default, override with TS

File: index.js [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+    },
+  ],
+  rules: {
+    "@typescript-eslint/explicit-function-return-type": "off",
+    "@typescript-eslint/no-unused-vars": [
+      "error",
+      {
+        varsIgnorePattern: "^_",
```
Comment (Averethel) [inline index.js:51]: lets double check for vars

File: index.js [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+        selector: "typeLike",
+        format: ["PascalCase"],
+      },
+    ],
+    "@typescript-eslint/no-use-before-define": [
+      "error",
+      { functions: false, variables: false },
+    ],
+    "prettier/prettier": ["error", { semi: false }],
+    "prefer-spread": "off",
```
Comment (Averethel) [inline index.js:104]: double-check if we can enable it

File: index.js [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+    "no-console": ["error", { allow: ["info", "warn", "error"] }],
```
Comment (Averethel) [inline index.js:105]: let's rever to disabling every console statement

File: index.js [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+    "no-irregular-whitespace": [
+      "error",
+      {
```
Comment (Averethel) [inline index.js:108]: apply only to json/move to locale config in nextjs config

File: index.js [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+        skipStrings: true,
+        skipTemplates: true,
+      },
+    ],
+    "sort-keys": "off",
+    "json/*": ["error", "allowComments"],
```
Comment (Averethel) [inline index.js:114]: dobule-check if allowComments is needed

File: react.js [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+        "jest/valid-expect-in-promise": "off",
+        "jest/valid-expect": "off",
+        "jest/no-standalone-expect": "off",
+        "testing-library/await-async-query": "off",
+      },
+    },
+    {
+      files: ["*.jsx"],
+      rules: {
+        "@typescript-eslint/no-var-requires": "off",
```
Comment (Averethel) [inline react.js:40]: this can go away with parser change

File: next.js [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+        ],
+      },
+    ],
+    "no-restricted-imports": [
+      "error",
+      {
+        paths: [
+          {
+            name: "next/link",
+            message: 'Please import from "~/components/appLink.tsx" instead.',
```
Comment (Averethel) [inline next.js:40]: try to message without the path

File: next.js [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+          },
+          {
+            name: "next/router",
+            importNames: ["default"],
+            message: 'Please import from "~/server/routing/routes.ts" instead.',
```
Comment (Averethel) [inline next.js:45]: try message without a path

File: index.js (hunk 28 above)
Comment (Averethel) [reply index.js:43]: Seems to work like a charm

File: index.js (hunk 29 above)
Comment (Averethel) [reply index.js:51]: Destructuring arrays is the main culprit here
```
const array = [1, 2, 3, 4]
```

We can write either:
```
const [ _one, two, _three, four ] = array
```

Or:
```
const [, two, , four] = array
```

I find the first much more readable and much less error-prone (imagine how easy it is to add an extra comma or remove one because you'd think there's too many.

Alternatively, we can disable the rule case by case basis, but this really opens up a possibility for a code that's quite hard to understand.

File: index.js (hunk 30 above)
Comment (Averethel) [reply index.js:104]: We violate this to build arrays of specific length via:
```Array.apply(value, { length: x })```
Should be fine to enable and deal with the violations
